File: tree/145/145.js
// solution 1
var postorderTraversal = function(root) {
  let result = [];
  const traverse = (root, result) => {
    if (!root) return;
    traverse(root.left, result);
    traverse(root.right, result);
    result.push(root.val);
  };
  traverse(root, result);
  return result;
};

// solution 2
var postorderTraversal = function(root) {
  let result = [];
  let stack = [];
  if (root) stack.push(root);
  while (stack.length !== 0) {
    const curr = stack.pop();
    result.push(curr.val);
    if (curr.left) stack.push(curr.left);
    if (curr.right) stack.push(curr.right);
  }
  return result.reverse();
};


//solution 3
var postorderTraversal = function (root) {
    let result = [];
    let stack = [];
    if (root) stack.push([root, false]);
    while (stack.length !== 0) {
        const [curr, visited] = stack.pop();
        if (visited) {
            result.push(curr.val);
        } else {
            stack.push([curr, true]);
            if (curr.right) stack.push([curr.right, false]);
            if (curr.left) stack.push([curr.left, false]);
        }

    }
    return result;
};

File: tree/145/145.py
# solution 1
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        self.traverse(root, result)
        return result

    def traverse(self, root, result):
        if root is None:
            return
        self.traverse(root.left, result)
        self.traverse(root.right, result)
        result.append(root.val)


# solution 2
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        stack = []
        if root:
            stack.append(root)
        while stack:
            curr = stack.pop()
            result.append(curr.val)
            if curr.left:
                stack.append(curr.left)
            if curr.right:
                stack.append(curr.right)
        return result[::-1]

#solution 3
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        result=[]
        stack=[]
        if root: stack.append((root,False))
        while stack:
            curr,visited=stack.pop()
            if visited:
                result.append(curr.val)
            else:
                stack.append((curr,True))
                if curr.right: stack.append((curr.right,False))
                if curr.left: stack.append((curr.left,False))
        return result
